# <font color="red"> MBA em IA e Big Data</font>
## <span style="color:red">Linguagens e Ferramentas para Inteligência Artificial e Big Data (Python e SQL)</span>

### <span style="color:darkred">Pandas: parte 2</span>

*Moacir Antonelli Ponti*<br>
*ICMC/USP São Carlos*

*(com materiais do Prof. Luis Gustavo Nonato)*

__Conteúdo:__

- Series
- Estatísticas
- Filtragem Agrupada
- Limpeza de Dados
- Preparação de Dados

__Referências__ <br>
- McKinney, W. and PyData Development Team [Pandas: powerful Python data analysis toolkit](https://pandas.pydata.org/pandas-docs/stable/pandas.pdf)
- Documentação Pandas [http://pandas.pydata.org/pandas-docs/stable/index.html](http://pandas.pydata.org/pandas-docs/stable/index.html)
- Wickham, H. Tidy data: https://vita.had.co.nz/papers/tidy-data.pdf 

In [4]:
import pandas as pd
import numpy as np

# Series

> Array unidimensional com um determinado tipo de dado

É similar a um dicionário, possuindo índice (que pode funcionar como chave) e valores.

Uma série pode ser obtida a partir de uma *coluna* de um dataframe

Pode ser criada por meio de um array numpy, um dicionário ou até um único escalar

In [2]:
serie_aleat = pd.Series(np.random.rand(8))
print(serie_aleat)

0    0.360171
1    0.774329
2    0.128181
3    0.074965
4    0.453658
5    0.683874
6    0.732500
7    0.793070
dtype: float64


Note que o índice dessa série, como não foi especificado, é numérico iniciando em 0

In [2]:
veic = {
    'AAA0A00': 1980,
    'BBB1B11': 2001,
    'CCC2C22': 1984,
    'DDD3D33': 2010,
    'EEE13E4': 2011}

In [5]:
s_veic = pd.Series(veic)
print(s_veic)

AAA0A00    1980
BBB1B11    2001
CCC2C22    1984
DDD3D33    2010
EEE13E4    2011
dtype: int64


Séries se comportam de forma muito similar à `ndarrays`

In [6]:
s_veic.mean()

1997.2

In [9]:
s_veic > s_veic.mean()

AAA0A00    False
BBB1B11     True
CCC2C22    False
DDD3D33     True
EEE13E4     True
dtype: bool

In [8]:
s_veic[s_veic > s_veic.mean()]

BBB1B11    2001
DDD3D33    2010
EEE13E4    2011
dtype: int64

In [9]:
s_veic.dtype

dtype('int64')

É possível obter um array de fato pelo atributo `array`

In [10]:
s_veic.array

<PandasArray>
[1980, 2001, 1984, 2010, 2011]
Length: 5, dtype: int64

Ou um numpy array com `to_numpy()`:

In [11]:
s_veic.to_numpy()

array([1980, 2001, 1984, 2010, 2011])

--- 



In [7]:
import pandas as pd
df = pd.read_csv('countries_data.tsv', sep='\t') 

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1704 entries, 0 to 1703
Data columns (total 6 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   country    1704 non-null   object 
 1   continent  1704 non-null   object 
 2   year       1704 non-null   int64  
 3   lifeExp    1704 non-null   float64
 4   pop        1704 non-null   int64  
 5   gdpPercap  1704 non-null   float64
dtypes: float64(2), int64(2), object(2)
memory usage: 80.0+ KB


In [16]:
df.sample(8)

country continent  year  lifeExp       pop     gdpPercap
383       Croatia    Europe  2007   75.748   4493312  14619.222720
778         Italy    Europe  2002   80.240  57926999  27968.098170
52      Argentina  Americas  1972   67.065  24779799   9443.038526
1568      Tunisia    Africa  1992   70.001   8523077   4332.720164
1503       Taiwan      Asia  1967   67.500  13648692   2643.858681
1653      Vietnam      Asia  1997   70.672  76048996   1385.896769
1214  Philippines      Asia  1962   54.757  30325264   1649.552153
26        Algeria    Africa  1962   48.303  11000948   2550.816880

### Relembrando

Vamos recuperar os dados: ano, população e expectativa de vida
* em Porto Rico
* a partir de 1990

In [15]:
# recuperando as colunas
df[['year','pop','lifeExp']]

year       pop  lifeExp
0     1952   8425333   28.801
1     1957   9240934   30.332
2     1962  10267083   31.997
3     1967  11537966   34.020
4     1972  13079460   36.088
...    ...       ...      ...
1699  1987   9216418   62.351
1700  1992  10704340   60.377
1701  1997  11404948   46.809
1702  2002  11926563   39.989
1703  2007  12311143   43.487

[1704 rows x 3 columns]

In [17]:
df.loc[ (df['country']=='Puerto Rico') & (df['year']>=1990) , ['year','pop','lifeExp']]

year      pop  lifeExp
1256  1992  3585176   73.911
1257  1997  3759430   74.917
1258  2002  3859606   77.778
1259  2007  3942491   78.746

Em numpy e pandas os operadores lógicos E, OU e NÃO : `&`, `|`, `~`

In [18]:
yearonly = df['year']
type(yearonly)
yearonly

0       1952
1       1957
2       1962
3       1967
4       1972
        ... 
1699    1987
1700    1992
1701    1997
1702    2002
1703    2007
Name: year, Length: 1704, dtype: int64

# Estatísticas

Há várias funções para análise dos dados.

Consultar a documentação oficial é importante (há muito ruído na internet/vídeos)

Básicas, que retornam um escalar como: `mean`, `std`, `quantile`, `min`, `max`

E mais complexas como: `cov` (covariance), `corr` (correlation) 

`describe` gera estatística descritiva para todas as colunas numéricas (ou série)

In [19]:
df['pop'].mean()

29601212.324530516

In [20]:
df.describe()

year      lifeExp           pop      gdpPercap
count  1704.00000  1704.000000  1.704000e+03    1704.000000
mean   1979.50000    59.474439  2.960121e+07    7215.327081
std      17.26533    12.917107  1.061579e+08    9857.454543
min    1952.00000    23.599000  6.001100e+04     241.165876
25%    1965.75000    48.198000  2.793664e+06    1202.060309
50%    1979.50000    60.712500  7.023596e+06    3531.846988
75%    1993.25000    70.845500  1.958522e+07    9325.462346
max    2007.00000    82.603000  1.318683e+09  113523.132900

In [21]:
df[['lifeExp', 'pop', 'gdpPercap']].corr()

lifeExp       pop  gdpPercap
lifeExp    1.000000  0.064955   0.583706
pop        0.064955  1.000000  -0.025600
gdpPercap  0.583706 -0.025600   1.000000

---

# `groupby()` e `agg()`

Alguns cálculos são interessantes de se fazer agrupando por certos valores

Nesse dataset, uma pergunta possível seria:
* Qual a expectativa de vida média, por ano, considerando todos os países?

In [22]:
df.groupby(['year'])

In [25]:
df.groupby(['year'])['lifeExp'].mean()

year
1952    49.057620
1957    51.507401
1962    53.609249
1967    55.678290
1972    57.647386
1977    59.570157
1982    61.533197
1987    63.212613
1992    64.160338
1997    65.014676
2002    65.694923
2007    67.007423
Name: lifeExp, dtype: float64

Pandas tem várias funções embutidas, mas pode ser que queiramos aplicar nossas próprias funções ou de biblioteca externa, como do `numpy`

Nesse caso, usamos `agg()` para aplicar a função desejada, agregada aos grupos

In [28]:
import numpy as np

df.groupby(['year'])['lifeExp'].agg(np.max)

year
1952    72.670
1957    73.470
1962    73.680
1967    74.160
1972    74.720
1977    76.110
1982    77.110
1987    78.670
1992    79.360
1997    80.690
2002    82.000
2007    82.603
Name: lifeExp, dtype: float64

E até mesmo gerar um novo dataframe, agregando múltiplas funções
* lista de tuplas contendo (nome,funcao)

In [30]:
df.groupby(['year'])['pop'].agg( [ 
                        ('media', np.mean), 
                        ('log10_somapop', lambda val: np.log2(val.sum())) ])

media  log10_somapop
year                             
1952  1.695040e+07      31.164563
1957  1.876341e+07      31.311166
1962  2.042101e+07      31.433298
1967  2.265830e+07      31.583283
1972  2.518998e+07      31.736094
1977  2.767638e+07      31.871899
1982  3.020730e+07      31.998141
1987  3.303857e+07      32.127395
1992  3.599092e+07      32.250877
1997  3.883947e+07      32.360767
2002  4.145759e+07      32.454880
2007  4.402122e+07      32.541443

É possível agrupar com múltiplas colunas

In [31]:
yearcont = df.groupby(['year','continent'])[['pop','gdpPercap']].agg([np.mean, np.std])
yearcont

pop                   gdpPercap              
                        mean           std          mean           std
year continent                                                        
1952 Africa     4.570010e+06  6.317450e+06   1252.572466    982.952116
     Americas   1.380610e+07  3.234163e+07   4079.062552   3001.727522
     Asia       4.228356e+07  1.132267e+08   5195.484004  18634.890865
     Europe     1.393736e+07  1.724745e+07   5661.057435   3114.060493
     Oceania    5.343003e+06  4.735083e+06  10298.085650    365.560078
1957 Africa     5.093033e+06  7.076042e+06   1385.236062   1134.508918
     Americas   1.547816e+07  3.553706e+07   4616.043733   3312.381083
     Asia       4.735699e+07  1.280961e+08   5787.732940  19506.515959
     Europe     1.459635e+07  1.783235e+07   6963.012816   3677.950146
     Oceania    5.970988e+06  5.291395e+06  11598.522455    917.644806
1962 Africa     5.702247e+06  7.957545e+06   1598.078825   1461.839189
     Americas   1.733081e+07  3.887683e+07   4901.541870   3421.740569
     Asia       5.140476e+07  1.361027e+08   5729.369625  16415.857196
     Europe     1.534517e+07  1.865642e+07   8365.486814   4199.193906
     Oceania    6.641759e+06  5.873524e+06  12696.452430    677.727301
1967 Africa     6.447875e+06  8.985505e+06   2050.363801   2847.717603
     Americas   1.922986e+07  4.192559e+07   5668.253496   4160.885560
     Asia       5.774736e+07  1.533418e+08   5971.173374  14062.591362
     Europe     1.603930e+07  1.944359e+07  10143.823757   4724.983889
     Oceania    7.300207e+06  6.465865e+06  14495.021790     43.986086
1972 Africa     7.305376e+06  1.013083e+07   2339.615674   3286.853884
     Americas   2.117537e+07  4.493546e+07   6491.334139   4754.404329
     Asia       6.518098e+07  1.740949e+08   8187.468699  19087.502918
     Europe     1.668784e+07  2.018034e+07  12479.575246   5509.691411
     Oceania    8.053050e+06  7.246360e+06  16417.333380    525.091980
1977 Africa     8.328097e+06  1.158518e+07   2585.938508   4142.398707
     Americas   2.312271e+07  4.790406e+07   7352.007126   5355.602518
     Asia       7.225799e+07  1.917074e+08   7791.314020  11815.777923
     Europe     1.723882e+07  2.056054e+07  14283.979110   5874.464896
     Oceania    8.619500e+06  7.713969e+06  17283.957605   1485.263517
1982 Africa     9.602857e+06  1.345624e+07   2481.592960   3242.632753
     Americas   2.521164e+07  5.129438e+07   7506.737088   5530.490471
     Asia       7.909502e+07  2.065415e+08   7434.135157   8701.176499
     Europe     1.770890e+07  2.097129e+07  15617.896551   6453.234827
     Oceania    9.197425e+06  8.466578e+06  18554.709840   1304.328377
1987 Africa     1.105450e+07  1.527748e+07   2282.668991   2566.531947
     Americas   2.731016e+07  5.445969e+07   7793.400261   6665.039509
     Asia       8.700669e+07  2.257332e+08   7608.226508   8090.262765
     Europe     1.810314e+07  2.136971e+07  17214.310727   7482.957960
     Oceania    9.787208e+06  9.150020e+06  20448.040160   2037.668013
1992 Africa     1.267464e+07  1.756272e+07   2281.810333   2644.075602
     Americas   2.957096e+07  5.810922e+07   8044.934406   7047.089191
     Asia       9.494825e+07  2.449604e+08   8639.690248   9727.431088
     Europe     1.860476e+07  2.212674e+07  17061.568084   9109.804361
     Oceania    1.045983e+07  9.930822e+06  20894.045885   3578.979883
1997 Africa     1.430448e+07  1.987301e+07   2378.759555   2820.728117
     Americas   3.187602e+07  6.203282e+07   8889.300863   7874.225145
     Asia       1.025238e+08  2.623497e+08   9834.093295  11094.180481
     Europe     1.896480e+07  2.274815e+07  19076.781802  10065.457716
     Oceania    1.112072e+07  1.052815e+07  24024.175170   4205.533703
2002 Africa     1.603315e+07  2.230300e+07   2599.385159   2972.651308
     Americas   3.399091e+07  6.560155e+07   9287.677107   8895.817785
     Asia       1.091455e+08  2.767017e+08  10174.090397  11150.719203
     Europe     1.927413e+07  2.322369e+07  21711

Nesse caso o índice será composto, tornando mais difícil acesso

Tente mostrar o `yearcont.index` para ver

In [32]:
yearcont.index

MultiIndex([(1952,   'Africa'),
            (1952, 'Americas'),
            (1952,     'Asia'),
            (1952,   'Europe'),
            (1952,  'Oceania'),
            (1957,   'Africa'),
            (1957, 'Americas'),
            (1957,     'Asia'),
            (1957,   'Europe'),
            (1957,  'Oceania'),
            (1962,   'Africa'),
            (1962, 'Americas'),
            (1962,     'Asia'),
            (1962,   'Europe'),
            (1962,  'Oceania'),
            (1967,   'Africa'),
            (1967, 'Americas'),
            (1967,     'Asia'),
            (1967,   'Europe'),
            (1967,  'Oceania'),
            (1972,   'Africa'),
            (1972, 'Americas'),
            (1972,     'Asia'),
            (1972,   'Europe'),
            (1972,  'Oceania'),
            (1977,   'Africa'),
            (1977, 'Americas'),
            (1977,     'Asia'),
            (1977,   'Europe'),
            (1977,  'Oceania'),
            (1982,   'Africa'),
        

Para facilitar, é útil "achatar" os índices, resetando-os

In [33]:
yearcont = yearcont.reset_index()
yearcont.head(10)

year continent           pop                   gdpPercap              
                           mean           std          mean           std
0  1952    Africa  4.570010e+06  6.317450e+06   1252.572466    982.952116
1  1952  Americas  1.380610e+07  3.234163e+07   4079.062552   3001.727522
2  1952      Asia  4.228356e+07  1.132267e+08   5195.484004  18634.890865
3  1952    Europe  1.393736e+07  1.724745e+07   5661.057435   3114.060493
4  1952   Oceania  5.343003e+06  4.735083e+06  10298.085650    365.560078
5  1957    Africa  5.093033e+06  7.076042e+06   1385.236062   1134.508918
6  1957  Americas  1.547816e+07  3.553706e+07   4616.043733   3312.381083
7  1957      Asia  4.735699e+07  1.280961e+08   5787.732940  19506.515959
8  1957    Europe  1.459635e+07  1.783235e+07   6963.012816   3677.950146
9  1957   Oceania  5.970988e+06  5.291395e+06  11598.522455    917.644806

---

#### <font color="blue">Exercício 2.9</font>

Carregue o arquivo `tips.csv`:
1. Filtre as linhas selecionando apenas jantares (time = 'Dinner') e cuja conta foi superior a 40 (total_bill > 40), mostrando o total da conta, número de pessoas na mesa e gorjeta (total_bill, size, tip);
2. Obtenha um novo dataframe em que seja mostrada a gorjeta (tip) média e máxima para cada valor de dia da semana (day) e horário (time)

In [44]:
# Resultados - 24/07/2022

dtips = pd.read_csv('data/tips.csv')
dtips.loc[(dtips["time"] == "Dinner") & (dtips["total_bill"] > 40), ["total_bill", "size", "tip"]]

newdf = dtips.groupby(["day", "time"])["tip"].agg([np.mean, np.max])

print(newdf)


                 mean   amax
day  time                   
Fri  Dinner  2.940000   4.73
     Lunch   2.282500   3.48
Sat  Dinner  3.019647  10.00
Sun  Dinner  3.255132   6.50
Thur Dinner  3.000000   3.00
     Lunch   2.780500   6.70


---
# Transformações com `apply`, `map` e `applymap`

* `apply` aplica uma função ao longo de linhas (axis=0) ou colunas (axis=1) de um DataFrame
* `map` aplica uma função em cada elemento de uma Serie
* `applymap` aplica uma função em cada elemento de um DataFrame

In [35]:
dtips

total_bill   tip     sex smoker   day    time  size
0         16.99  1.01  Female     No   Sun  Dinner     2
1         10.34  1.66    Male     No   Sun  Dinner     3
2         21.01  3.50    Male     No   Sun  Dinner     3
3         23.68  3.31    Male     No   Sun  Dinner     2
4         24.59  3.61  Female     No   Sun  Dinner     4
..          ...   ...     ...    ...   ...     ...   ...
240       29.03  5.92    Male     No   Sat  Dinner     3
241       27.18   NaN  Female    Yes   Sat  Dinner     2
242       22.67   NaN    Male    Yes   Sat  Dinner     2
243       17.82  1.75    Male     No   Sat  Dinner     2
244         NaN  3.00  Female     No  Thur  Dinner     2

[245 rows x 7 columns]

In [36]:
# soma das contas e gorjetas ao longo das linhas
dtips[['total_bill','tip']].apply(np.sum)

total_bill    4817.15
tip            727.43
dtype: float64

In [37]:
# soma da conta e gorjeta ao longo das colunas (total do dia)
dtips[['total_bill','tip']].apply(np.sum, axis=1)

0      18.00
1      12.00
2      24.51
3      26.99
4      28.20
       ...  
240    34.95
241    27.18
242    22.67
243    19.57
244     3.00
Length: 245, dtype: float64

In [38]:
# criando coluna com gorjeta média por pessoa na mesa
dtips['tip_perperson'] = dtips.apply(lambda x: np.round(x['tip']/x['size'],2), axis=1)
dtips

total_bill   tip     sex smoker   day    time  size  tip_perperson
0         16.99  1.01  Female     No   Sun  Dinner     2           0.50
1         10.34  1.66    Male     No   Sun  Dinner     3           0.55
2         21.01  3.50    Male     No   Sun  Dinner     3           1.17
3         23.68  3.31    Male     No   Sun  Dinner     2           1.66
4         24.59  3.61  Female     No   Sun  Dinner     4           0.90
..          ...   ...     ...    ...   ...     ...   ...            ...
240       29.03  5.92    Male     No   Sat  Dinner     3           1.97
241       27.18   NaN  Female    Yes   Sat  Dinner     2            NaN
242       22.67   NaN    Male    Yes   Sat  Dinner     2            NaN
243       17.82  1.75    Male     No   Sat  Dinner     2           0.88
244         NaN  3.00  Female     No  Thur  Dinner     2           1.50

[245 rows x 8 columns]

In [39]:
dtips[['total_bill','tip']].applymap(np.round)

total_bill  tip
0          17.0  1.0
1          10.0  2.0
2          21.0  4.0
3          24.0  3.0
4          25.0  4.0
..          ...  ...
240        29.0  6.0
241        27.0  NaN
242        23.0  NaN
243        18.0  2.0
244         NaN  3.0

[245 rows x 2 columns]

In [40]:
dtips['total_bill'].map(np.round)

0      17.0
1      10.0
2      21.0
3      24.0
4      25.0
       ... 
240    29.0
241    27.0
242    23.0
243    18.0
244     NaN
Name: total_bill, Length: 245, dtype: float64

---
# Limpeza de dados

## Tratando dados faltantes e duplicados

Esse dataset traz valores **faltantes**

Por padrão, esses serão carregados como sendo valores `NaN` do `numpy`

É preciso investigá-los pois eles são apenas ignorados pela maioria das funções

In [67]:
dtips = pd.read_csv('tips.csv')

pd.isna(dtips)

np.any(pd.isna(dtips))

True

### Percorrendo colunas

Por padrão a iteração é por linhas, vamos iterar pelas colunas procurando onde há dados faltantes

In [68]:
print('Dados faltantes em:')
for (colName, colData) in dtips.iteritems():
    if (np.any(pd.isna(colData))):
        print(colName)

Dados faltantes em:
total_bill
tip


Para filtrar especificamente os valores `NaN` há algumas opções, em particular `isna()`  combinado com `any(axis=1)` para verificar todas as colunas

In [69]:
np.any(dtips.isna(),axis=1)

0      False
1      False
2      False
3      False
4      False
       ...  
240    False
241     True
242     True
243    False
244     True
Length: 245, dtype: bool

In [70]:
dtips.loc[np.any(dtips.isna(),axis=1)]

total_bill  tip     sex smoker   day    time  size
238       32.83  NaN    Male    Yes   Sat  Dinner     2
241       27.18  NaN  Female    Yes   Sat  Dinner     2
242       22.67  NaN    Male    Yes   Sat  Dinner     2
244         NaN  3.0  Female     No  Thur  Dinner     2

### `dropna`

Remove todas as linhas ou colunas onde dados faltantes aparecem.

* `axis` define se linhas (padrão, 0) ou colunas (1)

In [71]:
print(dtips.shape)

(245, 7)


In [72]:
dtips_na_rows = dtips.dropna()
print(dtips_na_rows.shape)

(241, 7)


In [73]:
dtips_na_cols = dtips.dropna(axis=1)
print(dtips_na_cols.shape)

(245, 5)


### `fillna`

Substituir dados faltantes por algum valor

In [74]:
dtips.loc[np.any(dtips.isna(),axis=1)]

total_bill  tip     sex smoker   day    time  size
238       32.83  NaN    Male    Yes   Sat  Dinner     2
241       27.18  NaN  Female    Yes   Sat  Dinner     2
242       22.67  NaN    Male    Yes   Sat  Dinner     2
244         NaN  3.0  Female     No  Thur  Dinner     2

In [75]:
dtips_fill1 = dtips.fillna(0)

# mostrando dados preenchidos, filtrando pelos do dataframe com nans
dtips_fill1.loc[np.any(dtips.isna(),axis=1)]

total_bill  tip     sex smoker   day    time  size
238       32.83  0.0    Male    Yes   Sat  Dinner     2
241       27.18  0.0  Female    Yes   Sat  Dinner     2
242       22.67  0.0    Male    Yes   Sat  Dinner     2
244        0.00  3.0  Female     No  Thur  Dinner     2

Preenchimento diferenciado por colunas pode ser feito com dicionário

In [76]:
filldic = {'total_bill':-1, 'tip': 0}

dtips_fill2 = dtips.fillna(filldic)
dtips_fill2.loc[np.any(dtips.isna(),axis=1)]

total_bill  tip     sex smoker   day    time  size
238       32.83  0.0    Male    Yes   Sat  Dinner     2
241       27.18  0.0  Female    Yes   Sat  Dinner     2
242       22.67  0.0    Male    Yes   Sat  Dinner     2
244       -1.00  3.0  Female     No  Thur  Dinner     2

### `duplicated`, `drop_duplicates()`

Lida com linhas duplicados retorando `True` se detectada duplicação. 

In [77]:
print(dtips.duplicated())
print('\n',dtips.duplicated().any())

0      False
1      False
2      False
3      False
4      False
       ...  
240    False
241    False
242    False
243    False
244    False
Length: 245, dtype: bool

 True


- parâmetro `keep`:
    - False: retorna `True` para todas as duplicatas
    - first: a primeira não é detectada
    - last: a última não é detectada    

In [78]:
dtips[dtips.duplicated(keep=False)]

total_bill   tip   sex smoker  day    time  size
217        7.74  1.44  Male    Yes  Sat  Dinner     2
218        7.74  1.44  Male    Yes  Sat  Dinner     2
224       13.42  1.58  Male    Yes  Fri   Lunch     2
225       13.42  1.58  Male    Yes  Fri   Lunch     2

In [79]:
dtips[dtips.duplicated(keep='first')]

total_bill   tip   sex smoker  day    time  size
218        7.74  1.44  Male    Yes  Sat  Dinner     2
225       13.42  1.58  Male    Yes  Fri   Lunch     2

In [80]:
dtips[dtips.duplicated(keep='last')]

total_bill   tip   sex smoker  day    time  size
217        7.74  1.44  Male    Yes  Sat  Dinner     2
224       13.42  1.58  Male    Yes  Fri   Lunch     2

Para remover usamos `drop_duplicates()` valendo o mesmo parâmetro `keep`

In [81]:
dtips_first = dtips.drop_duplicates(keep='first')
dtips_last = dtips.drop_duplicates(keep='last')

# notar que foram removidos índices diferentes!
print(dtips_first.iloc[217:225])
print(dtips_last.iloc[217:225])

     total_bill   tip     sex smoker  day    time  size
217        7.74  1.44    Male    Yes  Sat  Dinner     2
219       30.14  3.09  Female    Yes  Sat  Dinner     4
220       12.16  2.20    Male    Yes  Fri   Lunch     2
221       13.42  3.48  Female    Yes  Fri   Lunch     2
222        8.58  1.92    Male    Yes  Fri   Lunch     1
223       15.98  3.00  Female     No  Fri   Lunch     3
224       13.42  1.58    Male    Yes  Fri   Lunch     2
226       16.27  2.50  Female    Yes  Fri   Lunch     2
     total_bill   tip     sex smoker  day    time  size
218        7.74  1.44    Male    Yes  Sat  Dinner     2
219       30.14  3.09  Female    Yes  Sat  Dinner     4
220       12.16  2.20    Male    Yes  Fri   Lunch     2
221       13.42  3.48  Female    Yes  Fri   Lunch     2
222        8.58  1.92    Male    Yes  Fri   Lunch     1
223       15.98  3.00  Female     No  Fri   Lunch     3
225       13.42  1.58    Male    Yes  Fri   Lunch     2
226       16.27  2.50  Female    Yes  Fri   Lunc

In [82]:
# nesses casos pode ser útil resetar os índices novamente!
dtips_first = dtips_first.reset_index()

In [83]:
print(dtips_first.iloc[217:225])

     index  total_bill   tip     sex smoker  day    time  size
217    217        7.74  1.44    Male    Yes  Sat  Dinner     2
218    219       30.14  3.09  Female    Yes  Sat  Dinner     4
219    220       12.16  2.20    Male    Yes  Fri   Lunch     2
220    221       13.42  3.48  Female    Yes  Fri   Lunch     2
221    222        8.58  1.92    Male    Yes  Fri   Lunch     1
222    223       15.98  3.00  Female     No  Fri   Lunch     3
223    224       13.42  1.58    Male    Yes  Fri   Lunch     2
224    226       16.27  2.50  Female    Yes  Fri   Lunch     2


---

# Preparação de dados

1. Cada variável forma uma coluna
2. Cada observação forma uma linha
3. Cada tipo de unidade observacional forma uma tabela

* 3.a forma normal de Codd!

**Exemplo**:

Recebemos duas tabelas, cada uma com temperaturas coletadas em cidades. A primeira é referente aos primeiros dias de coleta, a segunda aos dias seguintes, organizadas por outra pessoa e enviada.

Precisamos preparar os dados de forma que uma observação seja:
* uma cidade, um dia e uma temperatura naquele dia.

In [84]:
# criando os dataframes para o exemplo
temps = np.array([[23, 21, 20],[30, 29, 28],[18, 21, 20],[9, 10, 13]])
dtemp1 = pd.DataFrame(temps, index=['São Paulo','Fortaleza','Montevideo','London'])

dtemp2 = pd.DataFrame({
    'São Paulo': [25, 27, 23, 25],
    'Fortaleza': [35, 32, 31, 29],
    'Montevideo': [20, 18, 21, 23],
    'London': [14, 15, 12, 13]
})

In [85]:
dtemp1

0   1   2
São Paulo   23  21  20
Fortaleza   30  29  28
Montevideo  18  21  20
London       9  10  13

In [86]:
dtemp2

São Paulo  Fortaleza  Montevideo  London
0         25         35          20      14
1         27         32          18      15
2         23         31          21      12
3         25         29          23      13

### `melt()` para reconfigurar tabelas

Permite criar formato largo ou longo.

Auxilia muito para colocar o DataFrame de maneira que cada linha seja uma observação

* `id_vars` são as variáveis que não serão modificadas e continuarão como colunas
* `value_vars` são as variáveis que passarão a ser "valor" (linhas)

In [87]:
print(dtemp1)

print(dtemp1.melt())

             0   1   2
São Paulo   23  21  20
Fortaleza   30  29  28
Montevideo  18  21  20
London       9  10  13
    variable  value
0          0     23
1          0     30
2          0     18
3          0      9
4          1     21
5          1     29
6          1     21
7          1     10
8          2     20
9          2     28
10         2     20
11         2     13


In [88]:
print(dtemp2)
print(dtemp2.melt())

   São Paulo  Fortaleza  Montevideo  London
0         25         35          20      14
1         27         32          18      15
2         23         31          21      12
3         25         29          23      13
      variable  value
0    São Paulo     25
1    São Paulo     27
2    São Paulo     23
3    São Paulo     25
4    Fortaleza     35
5    Fortaleza     32
6    Fortaleza     31
7    Fortaleza     29
8   Montevideo     20
9   Montevideo     18
10  Montevideo     21
11  Montevideo     23
12      London     14
13      London     15
14      London     12
15      London     13


Gostaríamos de uma tabela no seguinte formato:

|city|day|temp|
|---|---|---|
|São Paulo|0|23|
|São Paulo|1|21|
|...|...|...|
|Londres|0|9|
|...|...|...|

Vamos começar pela primeira tabela e ver o que podemos fazer.

In [89]:
dtemp1.melt().head()

variable  value
0         0     23
1         0     30
2         0     18
3         0      9
4         1     21

As cidades somem pois essa informação está no índice.

In [90]:
dtemp1.index

Index(['São Paulo', 'Fortaleza', 'Montevideo', 'London'], dtype='object')

Solução: resetar o índice e criar uma nova coluna com os nomes das cidades

In [91]:
dtemp1 = dtemp1.rename_axis('city').reset_index()
dtemp1

city   0   1   2
0   São Paulo  23  21  20
1   Fortaleza  30  29  28
2  Montevideo  18  21  20
3      London   9  10  13

O uso automático 'derrete' toda a tabela, gostaríamos de **fixar a cidade** como uma variável identificadora 

In [94]:
dtemp1_melt = dtemp1.melt(id_vars=['city'])
dtemp1_melt

city variable  value
0    São Paulo        0     23
1    Fortaleza        0     30
2   Montevideo        0     18
3       London        0      9
4    São Paulo        1     21
5    Fortaleza        1     29
6   Montevideo        1     21
7       London        1     10
8    São Paulo        2     20
9    Fortaleza        2     28
10  Montevideo        2     20
11      London        2     13

Outro ponto que seria interessante é nomear as variáveis no formato longo

In [95]:
tabela1 = dtemp1.melt(id_vars=['city'], value_name='temperature', var_name='day')
tabela1

city day  temperature
0    São Paulo   0           23
1    Fortaleza   0           30
2   Montevideo   0           18
3       London   0            9
4    São Paulo   1           21
5    Fortaleza   1           29
6   Montevideo   1           21
7       London   1           10
8    São Paulo   2           20
9    Fortaleza   2           28
10  Montevideo   2           20
11      London   2           13

Agora na tabela 2.

O `melt` aqui já está mais próximo, veja que nesse caso as colunas informam o nome das cidades, então perdemos o valor dos dias

In [96]:
print(dtemp2)
print(dtemp2.melt())

   São Paulo  Fortaleza  Montevideo  London
0         25         35          20      14
1         27         32          18      15
2         23         31          21      12
3         25         29          23      13
      variable  value
0    São Paulo     25
1    São Paulo     27
2    São Paulo     23
3    São Paulo     25
4    Fortaleza     35
5    Fortaleza     32
6    Fortaleza     31
7    Fortaleza     29
8   Montevideo     20
9   Montevideo     18
10  Montevideo     21
11  Montevideo     23
12      London     14
13      London     15
14      London     12
15      London     13


Assim, podemos seguir o que foi feito na primeira tabela "transpondo" o DataFrame

In [97]:
dtemp2.transpose()

0   1   2   3
São Paulo   25  27  23  25
Fortaleza   35  32  31  29
Montevideo  20  18  21  23
London      14  15  12  13

In [98]:
dtemp2_tran = dtemp2.transpose().rename_axis('city').reset_index()
dtemp2_tran

city   0   1   2   3
0   São Paulo  25  27  23  25
1   Fortaleza  35  32  31  29
2  Montevideo  20  18  21  23
3      London  14  15  12  13

In [99]:
tabela2 = dtemp2_tran.melt(id_vars=['city'],  value_name='temperature', var_name='day')
tabela2

city day  temperature
0    São Paulo   0           25
1    Fortaleza   0           35
2   Montevideo   0           20
3       London   0           14
4    São Paulo   1           27
5    Fortaleza   1           32
6   Montevideo   1           18
7       London   1           15
8    São Paulo   2           23
9    Fortaleza   2           31
10  Montevideo   2           21
11      London   2           12
12   São Paulo   3           25
13   Fortaleza   3           29
14  Montevideo   3           23
15      London   3           13

No entanto, veja que os dias da tabela 2 começam em 0

Sabendo que os dias da tabela 2 vem logo após os dias da tabela 1, podemos:
1. pegar o último dia na tabela 1
2. adicionar esse valor ao cada dia da tabela 2 usando `map` ou `apply`

In [100]:
last_day = tabela1['day'].max()

tabela2['day'] = tabela2['day'].apply(lambda x: x+last_day+1)
tabela2

city  day  temperature
0    São Paulo    3           25
1    Fortaleza    3           35
2   Montevideo    3           20
3       London    3           14
4    São Paulo    4           27
5    Fortaleza    4           32
6   Montevideo    4           18
7       London    4           15
8    São Paulo    5           23
9    Fortaleza    5           31
10  Montevideo    5           21
11      London    5           12
12   São Paulo    6           25
13   Fortaleza    6           29
14  Montevideo    6           23
15      London    6           13

## Combinando

No caso anterior temos 2 DataFrames, mas desejamos combiná-los

- `merge`: combina dados vinculando linhas por uma ou mais chave
- `concatenate`: combina dados a partir de um eixo específico

### `merge`

Realiza a fusão de dois: DataFrames, Series, ou Serie com DataFrame.

Fusão com base em:
- valores de colunas, uma em cada DataFrame 
- valores de uma coluna em um dos DataFrame e índices no outro
- valores dos índices em ambos os DataFrames (ou série)

#### Merge baseado nos valores de colunas

In [101]:
# construindo dois DataFrames que serão fundidos com o método 'merge'
df1 = pd.DataFrame({'key': ['a','a','b','b','c','a'], 
                   'data1': range(10,70,10)})

df2 = pd.DataFrame({'key': ['b','a','a','b'], 
                   'data2': range(4)})

print(df1)
print(df2)

  key  data1
0   a     10
1   a     20
2   b     30
3   b     40
4   c     50
5   a     60
  key  data2
0   b      0
1   a      1
2   a      2
3   b      3


In [102]:
# fusão com base nos valores de 'key'
pd.merge(df1, df2, on='key')  
# Apenas os valores que aparencem em ambos são utilizados no 'merge'

key  data1  data2
0   a     10      1
1   a     10      2
2   a     20      1
3   a     20      2
4   a     60      1
5   a     60      2
6   b     30      0
7   b     30      3
8   b     40      0
9   b     40      3

Parâmetro `how` força que todos os valores sejam utilizados
* `outer`: realiza união entre os dados
* `inner`: realiza intersecção entre os dados

In [103]:
pd.merge(df1,df2,on='key',how='outer')  # o valor 'c' aparece na coluna 'key'
                                        # apenas em df1, enquanto o valor 'd'
                                        # aparece apenas em df2    

key  data1  data2
0    a     10    1.0
1    a     10    2.0
2    a     20    1.0
3    a     20    2.0
4    a     60    1.0
5    a     60    2.0
6    b     30    0.0
7    b     30    3.0
8    b     40    0.0
9    b     40    3.0
10   c     50    NaN

#### Merge baseado nos valores de coluna em DataFrame e rótulos no outro DataFrame
Para ser realizar o merging com base em colunas e rótulos, deve-se especificar os parâmetros
`left_on`, `right_on`, `left_index` e `right_index` de forma apropriada.
- __left\_on__ = nome da coluna do DataFrame da esquerda (primeiro parâmetro)
- __left\_index__ = True indica que o índice no DataFrame da esquerda (primeiro parâmetro) serão utilizados 
- __right\_on__ = nome da coluna do DataFrame da direita (segundo parâmetro)
- __right\_index__ = True indica que o índice no DataFrame da direita (segundo parâmetro) serão utilizados 

In [110]:
dcity = pd.DataFrame({'country':['BR','BR','UR']}, index=['São Paulo','Fortaleza','Montevideo'])
dcity

country
São Paulo       BR
Fortaleza       BR
Montevideo      UR

In [111]:
tabela1

city day  temperature
0    São Paulo   0           23
1    Fortaleza   0           30
2   Montevideo   0           18
3       London   0            9
4    São Paulo   1           21
5    Fortaleza   1           29
6   Montevideo   1           21
7       London   1           10
8    São Paulo   2           20
9    Fortaleza   2           28
10  Montevideo   2           20
11      London   2           13

In [114]:
# Merge utilizando a coluna "city" da tabela 1 (de temperaturas) e os índices em dcity
pd.merge(tabela1, dcity, left_on='city', right_index=True, how='outer')

city day  temperature country
0    São Paulo   0           23      BR
4    São Paulo   1           21      BR
8    São Paulo   2           20      BR
1    Fortaleza   0           30      BR
5    Fortaleza   1           29      BR
9    Fortaleza   2           28      BR
2   Montevideo   0           18      UR
6   Montevideo   1           21      UR
10  Montevideo   2           20      UR
3       London   0            9     NaN
7       London   1           10     NaN
11      London   2           13     NaN

Se não especificarmos união (`outer`), apenas ocorrências de valores em ambos será considerada

In [115]:
pd.merge(tabela1, dcity, left_on='city', right_index=True)

city day  temperature country
0    São Paulo   0           23      BR
4    São Paulo   1           21      BR
8    São Paulo   2           20      BR
1    Fortaleza   0           30      BR
5    Fortaleza   1           29      BR
9    Fortaleza   2           28      BR
2   Montevideo   0           18      UR
6   Montevideo   1           21      UR
10  Montevideo   2           20      UR

### `concat`

Concatena DataFrames e/ou Series nas direções:
- __axis = 0__ concatena por linhas 
- __axis = 1__ concatena por colunas. Quando empregado em um conjunto de séries resulta em um DataFrame

O método assume como parâmetro uma lista de DataFrames ou séries.

In [118]:
# Construindo 3 séries especificando os rótulos das linhas
s1 = pd.Series([0, 1], index=['a', 'b'])
s2 = pd.Series([2, 3, 4], index=['a', 'c', 'd'])
s3 = pd.Series([1, 5, 6], index=['a','b', 'e']) 
print(s1)
print(s2)
print(s3)

# Contatenando as 3 séries na vertical
# o resultado é uma nova série (rótulos podem estar repetidos)
serie1 = pd.concat([s1,s2,s3])

a    0
b    1
dtype: int64
a    2
c    3
d    4
dtype: int64
a    1
b    5
e    6
dtype: int64


In [119]:
serie1

a    0
b    1
a    2
c    3
d    4
a    1
b    5
e    6
dtype: int64

In [120]:
# concatenando as 3 séries por colunas especificando o parâmetro 'axis=1'
# o resutado é um DataFrame
print(pd.concat([s1,s2,s3],axis=1))

     0    1    2
a  0.0  2.0  1.0
b  1.0  NaN  5.0
c  NaN  3.0  NaN
d  NaN  4.0  NaN
e  NaN  NaN  6.0


In [121]:
print(pd.concat([tabela1,tabela2], axis=1))

          city  day  temperature        city  day  temperature
0    São Paulo    0         23.0   São Paulo    3           25
1    Fortaleza    0         30.0   Fortaleza    3           35
2   Montevideo    0         18.0  Montevideo    3           20
3       London    0          9.0      London    3           14
4    São Paulo    1         21.0   São Paulo    4           27
5    Fortaleza    1         29.0   Fortaleza    4           32
6   Montevideo    1         21.0  Montevideo    4           18
7       London    1         10.0      London    4           15
8    São Paulo    2         20.0   São Paulo    5           23
9    Fortaleza    2         28.0   Fortaleza    5           31
10  Montevideo    2         20.0  Montevideo    5           21
11      London    2         13.0      London    5           12
12         NaN  NaN          NaN   São Paulo    6           25
13         NaN  NaN          NaN   Fortaleza    6           29
14         NaN  NaN          NaN  Montevideo    6      

In [122]:
tabela_temperturas = pd.concat([tabela1,tabela2],axis=0)
print(tabela_temperturas)

          city day  temperature
0    São Paulo   0           23
1    Fortaleza   0           30
2   Montevideo   0           18
3       London   0            9
4    São Paulo   1           21
5    Fortaleza   1           29
6   Montevideo   1           21
7       London   1           10
8    São Paulo   2           20
9    Fortaleza   2           28
10  Montevideo   2           20
11      London   2           13
0    São Paulo   3           25
1    Fortaleza   3           35
2   Montevideo   3           20
3       London   3           14
4    São Paulo   4           27
5    Fortaleza   4           32
6   Montevideo   4           18
7       London   4           15
8    São Paulo   5           23
9    Fortaleza   5           31
10  Montevideo   5           21
11      London   5           12
12   São Paulo   6           25
13   Fortaleza   6           29
14  Montevideo   6           23
15      London   6           13


---

#### <font color="blue">Exercício 2.10</font>

Carregue os arquivos `sales1.csv`, `sales1_shipdate.csv`  e `sales2.csv`, os quais possuem informações de vendas realizadas. Devemos juntar as bases de dados e tratá-las. 

O arquivo `sales1_shipdate.csv` contém as datas de envio das ordens na `sales1.csv`. Já `sales2.csv` contém essa coluna no próprio arquivo

Para isso:

1. Combine as bases de dados, consolidando-as em um único DataFrame
2. Exiba na tela quais atributos possuem dados faltantes após a concatenação
    * Sabendo que `Total Revenue` é a multiplicação do preço unitário pela quantidade de unidades, preencha os valores faltantes dessa coluna
3. Detecte linhas duplicadas. Remova duplicatas, mantendo a primeira ocorrência, e imprima na tela quantas linhas foram removidas

In [55]:
df_sales1 = pd.read_csv("data/sales1.csv")
df_sales2 = pd.read_csv("data/sales2.csv")
df_shipdate = pd.read_csv("data/sales1_shipdate.csv")

# 1)
df_sales = pd.concat([df_sales1, df_sales2, df_shipdate], axis=1)

#2)
df_sales


Country        Item Type  Online Order Priority Order Date  \
0                  Tuvalu        Baby Food     0.0              H  5/28/2010   
1                  Brazil           Cereal     1.0              C  8/22/2012   
2                  Russia  Office Supplies     0.0              L   5/2/2014   
3   Sao Tome and Principe           Fruits     1.0              C  6/20/2014   
4                  Rwanda  Office Supplies     0.0              L   2/1/2013   
..                    ...              ...     ...            ...        ...   
83                    NaN              NaN     NaN            NaN        NaN   
84                    NaN              NaN     NaN            NaN        NaN   
85                    NaN              NaN     NaN            NaN        NaN   
86                    NaN              NaN     NaN            NaN        NaN   
87                    NaN              NaN     NaN            NaN        NaN   

       Order ID  Units Sold Unit Price  Total Cost  Total Profit  ...  \
0   669165933.0      9925.0     255.28  1582243.50     951410.50  ...   
1   963881480.0      2804.0     205.70   328376.44     248406.36  ...   
2   341417157.0      1779.0     651.21   933903.84     224598.75  ...   
3   514321792.0      8102.0       9.33    56065.84      19525.82  ...   
4   115456712.0      5062.0     651.21  2657347.52     639077.50  ...   
..          ...         ...        ...         ...           ...  ...   
83          NaN         NaN        NaN         NaN           NaN  ...   
84          NaN         NaN        NaN         NaN           NaN  ...   
85          NaN         NaN        NaN         NaN           NaN  ...   
86          NaN         NaN        NaN         NaN           NaN  ...   
87          NaN         NaN        NaN         NaN           NaN  ...   

    Order Date   Order ID  Units Sold Unit Price Total Revenue  Total Cost  \
0    1/13/2017  187310731        8263     109.28     902980.64   296145.92   
1     2/8/2017  522840487        8974     668.27    5997054.98  4509793.96   
2    2/19/2014  832401311        4901      81.73     400558.73   277739.67   
3    4/23/2012  972292029        1673     109.28     182825.44    59960.32   
4   11/19/2016  419123971        6952     437.20    3039414.40  1830670.16   
..         ...        ...         ...        ...           ...         ...   
83   7/26/2011  512878119         888     109.28      97040.64    31825.92   
84  11/11/2011  810711038        6267       9.33      58471.11    43367.64   
85    6/1/2016  728815257        1485     154.06     228779.10   135031.05   
86   7/30/2015  559427106        5767      81.73     471336.91   326815.89   
87   2/10/2012  665095412        5367     668.27    3586605.09  2697132.18   

    Total Profit   Ship Date     Order ID  Ship Date  
0      606834.72    3/1/2017  669165933.0  6/27/2010  
1     1487261.02   2/13/2017  963881480.0  9/15/2012  
2      122819.06   2/23/2014  341417157.0   5/8/2014  
3      122865.12    6/3/2012  514321792.0   7/5/2014  
4     1208744.24  12/18/2016  115456712.0   2/6/2013  
..           ...         ...          ...        ...  
83      65214.72    9/3/2011          NaN        NaN  
84      15103.47  12/28/2011          NaN        NaN  
85      93748.05   6/29/2016          NaN        NaN  
86     144521.02    8/8/2015          NaN        NaN  
87     889472.91   2/15/2012          NaN        NaN  

[88 rows x 24 columns]